### PR TITLE
Ports #10649 from Citadel: An attempt at making hijack not shit and more about point defence

### PR DESCRIPTION
**Pull request by necromanceranne**
## About The Pull Request

Read up here about the pr: https://github.com/Citadel-Station-13/Citadel-Station-13/pull/10649

It would be good to merge this pr with this one so as to prevent an obvious exploitable weakness to prevent hijacks. https://github.com/tgstation/tgstation/pull/54247

## Why It's Good For The Game

So essentially there was discussion already here (https://tgstation13.org/phpBB/viewtopic.php?f=33&t=27620) about how hijack is one of the few objectives on Manuel that you can murderbone with, which is mildly at odds with the intention of the server. 

Personally, I think hijack is just straight up terrible regardless of what server you're on. The way it functions is dreadful. Nobody must be aboard the shuttle in any part of the shuttle at all. This means you have people hiding in the walls or hiding in some invisible corner of the shuttle, cucking you of the hijack. For the most part, the only way to actually hijack the shuttle is to render it completely inhospitable or destroying it utterly save for, maybe, one square. The one you are standing on.

It's absurd. So, I knew kev made a pretty honest attempt at reworking hijack. And it works pretty well. It doesn't necessarily solve the problem of hijackers being mandatory mass shooters, but it goes some ways to improving the objective in a more interesting fashion, and allows for the discussion around the objective to be a little more open ended for the sake of the higher roleplay servers without actually detracting from the lower roleplay servers at all. If anything, this should improve the experience of being that gamer to stage a hijacking all the more interesting and about robustness, and less of hide-and-go-seek.

## Changelog
:cl: silicons
add: Shuttle hijacking has been completely reworked.
add: To hijack the shuttle, you must Alt-Click the emergency shuttle console (the one you emag) to begin the hijack process after a timed delay. You must be hijack capable to do this.
add: Each successful hack attempt (with a time based on antag role and whether or not you have the hijack objective) will push the hijack stage one step, up to 5.
add: After five successful hack attempts, the shuttle will be redirected. 
add: The antagonists who can hijack include: traitor, blood brothers, wizard, nuke ops, highlanders, wizards, changelings and heretics.
tweak: Preserves the old hijack as the new 'Elimination' objective. Nobody but highlanders have this, and their version accounts for only their solo hijacking.
/:cl:

### CODE REVIEW

File: code/modules/antagonists/official/official.dm [hunk 1]
```diff
@@ -4,7 +4,6 @@
 	show_in_antagpanel = FALSE
 	var/datum/objective/mission
 	var/datum/team/ert/ert_team
-	can_hijack = HIJACK_PREVENT
```
Comment (Fikou) [inline code/modules/antagonists/official/official.dm:7]: uhh

File: code/modules/antagonists/ert/ert.dm [hunk 2]
```diff
@@ -17,7 +17,6 @@
 	show_in_antagpanel = FALSE
 	show_to_ghosts = TRUE
 	antag_moodlet = /datum/mood_event/focused
-	can_hijack = HIJACK_PREVENT
```
Comment (Fikou) [inline code/modules/antagonists/ert/ert.dm:20]: eeee

File: code/modules/antagonists/official/official.dm (hunk 1 above)
Comment (necromanceranne) [reply code/modules/antagonists/official/official.dm:7]: I noticed, don't stress. I'll figure out some way to fix it, but it's approaching midnight so this is a problem for tomorrow me.

**Comment by Tharcoonvagh** [comment]
Shuttle as in the shuttle computer?

**Comment by necromanceranne** [comment]
You know the console you emag to immediately start departure? That console.

**Comment by necromanceranne** [comment]
Guess Travis wants to co-operate now.

File: code/modules/shuttle/emergency.dm [hunk 3]
```diff
@@ -152,6 +176,68 @@
...
+
+/obj/machinery/computer/emergency_shuttle/proc/announce_hijack_stage()
+	var/msg
+	switch(SSshuttle.emergency.hijack_status)
+		if(NOT_BEGUN)
+			return
+		if(STAGE_1)
+			var/datum/species/S = new
+			msg = "AUTHENTICATING - FAIL. AUTHENTICATING - FAIL. AUTHENTICATING - FAI###### Welcome, technician JOHN DOE."
+			qdel(S)
```
Comment (nightred) [inline code/modules/shuttle/emergency.dm:228]: Why are you creating a new species `S` only to qdel it 2 lines later?

**Comment by Tharcoonvagh** [comment]
> You know the console you emag to immediately start departure? That console.

Okay mr pr editor now people will think that Im dumb

File: code/modules/shuttle/emergency.dm (hunk 3 above)
Comment (necromanceranne) [reply code/modules/shuttle/emergency.dm:228]: I did nothing but that is fucking weird. I think it's something left in from a draft form of this. I believe that line originally checked for identity and sex.

**Comment by gitbirb** [comment]
make sure this doesn't make BYOS kit the meta shuttle against hijackers, since BYOS kit doesn't have a shuttle console last I recall

File: code/modules/antagonists/_common/antag_datum.dm [hunk 4]
```diff
@@ -264,6 +265,13 @@ GLOBAL_LIST_EMPTY(antagonists)
 		return
 	antag_memory = new_memo
 
+/// Gets how fast we can hijack the shuttle, return 0 for can not hijack. Defaults to hijack_speed var, override for custom stuff like buffing hijack speed for hijack objectives or something.
+/datum/antagonist/proc/hijack_speed()
+	var/datum/objective/hijack/H = locate() in objectives
+	if(!isnull(H?.hijack_speed_override))
+		return H.hijack_speed_override
+	return hijack_speed
```
Comment (Jared-Fogle) [inline code/modules/antagonists/_common/antag_datum.dm:273]:
```suggestion
	return H?.hijack_speed_override || hijack_speed
```

File: code/modules/shuttle/emergency.dm [hunk 5]
```diff
@@ -152,6 +176,68 @@
...
+		if(shuttle.hijack_status == HIJACKED)
+			shuttle.setTimer(hijack_completion_flight_time_set)
+		else
+			shuttle.setTimer(shuttle.timeLeft(1) + hijack_flight_time_increase)		//give the guy more time to hijack if it's already in flight.
+	return shuttle.hijack_status
+
+/obj/machinery/computer/emergency_shuttle/AltClick(user)
+	attempt_hijack_stage(user)
+
+/obj/machinery/computer/emergency_shuttle/proc/attempt_hijack_stage(mob/living/user)
```
Comment (Jared-Fogle) [inline code/modules/shuttle/emergency.dm:196]: I believe you have to check `isliving` as well. Ghosts can still alt click.

File: code/modules/shuttle/emergency.dm [hunk 6]
```diff
@@ -4,15 +4,39 @@
...
 	icon_screen = "shuttle"
 	icon_keyboard = "tech_key"
-
+	resistance_flags = INDESTRUCTIBLE
 	var/auth_need = 3
 	var/list/authorized = list()
 	var/list/acted_recently = list()
+	var/hijack_last_stage_increase = 0
+	var/hijack_stage_time = 50
+	var/hijack_stage_cooldown = 50
```
Comment (Jared-Fogle) [inline code/modules/shuttle/emergency.dm:25]: Use the cooldown defines.

File: code/modules/shuttle/emergency.dm [hunk 7]
```diff
@@ -4,15 +4,39 @@
...
+	var/hijack_flight_time_increase = 300
+	var/hijack_completion_flight_time_set = 100		//How long in deciseconds to set shuttle's timer after hijack is done.
```
Comment (Jared-Fogle) [inline code/modules/shuttle/emergency.dm:27]: For this and everything else, use the SECONDS define.

File: code/modules/shuttle/emergency.dm [hunk 8]
```diff
@@ -152,6 +176,68 @@
...
+		if(STAGE_2)
+			msg = "Warning: Navigational route fails \"IS_AUTHORIZED\". Please try againNN[scramble_message_replace_chars("againagainagainagainagain", 70)]."
+		if(STAGE_3)
+			msg = "CRC mismatch at ~h~ in calculated route buffer. Full reset initiated of FTL_NAVIGATION_SERVICES. Memory decrypted for automatic repair."
+		if(STAGE_4)
+			msg = "~ACS_directive module_load(cyberdyne.exploit.nanotrasen.shuttlenav)... NT key mismatch. Confirm load? Y...###Reboot complete. $SET transponder_state = 0; System link initiated with connected engines..."
+		if(HIJACKED)
+			msg = "<font color='red'><b>SYSTEM OVERRIDE</b></font> - Resetting course to \[[scramble_message_replace_chars("###########", 100)]\] \
+			([scramble_message_replace_chars("#######", 100)]/[scramble_message_replace_chars("#######", 100)]/[scramble_message_replace_chars("#######", 100)]) \
+			{AUTH - ROOT (uid: 0)}.</font>[SSshuttle.emergency.mode == SHUTTLE_ESCAPE? "Diverting from existing route - Bluespace exit in [hijack_completion_flight_time_set/10] seconds." : ""]"
```
Comment (Jared-Fogle) [inline code/modules/shuttle/emergency.dm:238]:
```suggestion
			{AUTH - ROOT (uid: 0)}.</font>[SSshuttle.emergency.mode == SHUTTLE_ESCAPE ? "Diverting from existing route - Bluespace exit in [hijack_completion_flight_time_set/10] seconds." : ""]"
```

**Comment by itseasytosee** [comment]
This makes me want to add more hijack locations. Wizard ending up near the wizard ship, and heretics sending them violently crashing into a large nearby asteroid or something. Also, you'd think that someone with some amount of access would be able to correct the shuttle's course if given an equal amount of time, if not less for someone like the captain.

**Comment by Reed0506** [comment]
just give BYOS a console, easy fix.

**Comment by ATH1909** [comment]
IIRC, aren't emergency shuttle consoles NOT indestructible? It might become the new meta to just shotgun the emergency shuttle console as soon as you can as a crewmember to keep traitors from using it to hijack.

**Comment by zxaber** [comment]
The console should probably be indestructible or else the way you solve this is body the hijacker out of the way long enough for you to smash it (and also just decon the BYOS console anytime beforehand).

By the by, Highlanders now use a special hijack objective because I needed it to fail with more than one highlander (including borgs). Highlander should probably be left to the old system, since it fits the event more. It's an admin-only mass murderbone thing already.

**Comment by ATH1909** [comment]
Does alt+clicking break invisibility from the chameleon mutation?

**Comment by Jared-Fogle** [comment]
If you shotgun the shuttle console to stop traitors from hijacking, then that sounds like an administration problem and not a code one.

**Comment by necromanceranne** [comment]
> The console should probably be indestructible.

It is.

File: code/modules/shuttle/emergency.dm (hunk 5 above)
Comment (necromanceranne) [reply code/modules/shuttle/emergency.dm:196]: Ghost traitor hijacks the shuttle hehe

**Comment by saprasam** [comment]
just kill the entire station before hand : - D

**Comment by necromanceranne** [comment]
> just kill the entire station before hand : - D

Do whatever you want, you're an antag. That doesn't remove that as an option, it's just hopefully making the objective less badly designed in principle.

File: code/modules/antagonists/_common/antag_datum.dm [hunk 9]
```diff
@@ -270,7 +271,7 @@ GLOBAL_LIST_EMPTY(antagonists)
 	var/datum/objective/hijack/H = locate() in objectives
 	if(!isnull(H?.hijack_speed_override))
 		return H.hijack_speed_override
```
Comment (Jared-Fogle) [inline code/modules/antagonists/_common/antag_datum.dm]:
```suggestion
```

If you have my other suggestion, then this is redundant and should be removed.

**Comment by moocowswag** [comment]
My only potential problem with this is now it adds another requirement for hijacking (getting access to the front of the shuttle) which really isnt much at all but I know it could end up angering people who didnt realize this up until they were just stuck fumbling their thumbs after leaving without everyone.

Infact, since the wire hacking is split up between the type of doors, what type of door do the shuttle ones / front count as? Are they their own set that you have to learn after the shuttle docks or are they command + public doors that you can figure out earlier before the shuttle docks?

**Comment by EgorDinamit** [comment]
Imo old hijack mechanic should still work, to prevent crew metagaming.
It still should hijack if you murder everyone on the shuttle, but the idea of being able to do it without killing anyone sounds good to me, it just shouldn't be obligatory.

**Comment by necromanceranne** [comment]
I'd much prefer one or the other. If players are metagaming this to an extreme degree, either by forcing early calls or barricading in the cockpit, that should be dealt with administratively if it's a problem. Already the older version could be metagamed in the form of hiding aboard (or in the walls of) the shuttle to prevent the hijack.

This is also not meant to force the issue on whether you need to murderbone to get hijack. If anything this is just making that somewhat less prone to failure due to conditions largely out of your control and far too difficult to reasonably resolve, like someone quite literally invisible to the naked eye hiding in some backend corner of the shuttle or in the walls. You can much more easily achieve a hijack by seizing complete control of the cockpit, after all.

If players are metagaming this somehow to be antifun, however that might be achieved, it's an admin issue and if admins don't deal with it, that's something you take up with them in a policy thread. There is precedent already about antifun behaviours being bannable if done in excess.

**Comment by Learner10** [comment]
> make sure this doesn't make BYOS kit the meta shuttle against hijackers, since BYOS kit doesn't have a shuttle console last I recall. Reason this is important is mainly because it'd feel really annoying if someone was a hijack traitor and just hears the "BYOS Kit has just been purchased" announcement, resulting in losing all hope and probably going on a murder spree, which could even be a non-meta "fuck this traitor in particular" thing, instead just being random chance, since people just purchase the kit sometimes, and by such a small unstoppable chance you immediately lost.

The easier solution would be to create the components needed to construct a shuttle console, make those components indestructible, and then force tie those components to the shuttle area so that they can't be taken away from said shuttle (in similar fashion to attempting to take the nuclear disk off the z level). That way, it ties in way better with the BYOS kit, since you can just assemble the shuttle console anywhere without being arbitrarily forced to work around an unmovable console, and a clever hijacker could attempt to construct the BYOS kit so that the console is super secretly hidden.

**Comment by EgorDinamit** [comment]
Nothing about code issues should be left for admins to deal with.
**If possible**, it should be coded the way to prevent most administrative issues before-hand.

If you are going to leave it as the only way to hijack, it shouldn't be possible to metagame in an **easy** way.
Making shuttle console indestructible would fix the easiest way to prevent traitors from hijacking shuttle.

I didn't check the code, but I think it's better to add a config for the possibility to hijack the shuttle without objective, people on LRP might like to do so from time to time.

**Comment by necromanceranne** [comment]
An indestructible console in the cockpit is a difficult thing to prevent access too without either early launching simply for the sake of preventing any hijack attempt at all (which is bannable already without good justification like some impeding threat), or quickly constructing some massive barricade that can't be bypassed in the same time it takes for someone to even get into that room while you're constructing the barricade. Any methods I could think of to stop a hijack now are not any better than before.

Given the nature of our tools and constructible objects, and the slew of options a hijacker has to give them an edge, I have my doubts but I stand to be impressed. So I look forward to seeing what could be done. But if players are somehow going to extreme lengths to prevent a hijack with no justification, consider ahelping.

Any of the mainstay antags can hijack that shuttle the new way as reasonably as the old way, and the old way was significantly more prone to failure.

**Comment by Jared-Fogle** [comment]
> people on LRP might like to do so from time to time.

There's nothing stopping you from, you know, just killing everyone.

**Comment by TetraK1** [comment]
shouldn't there be a consequence for smashing the console as crew, e.g. landing imperfectly at centcomm and gibbing half the crew before round-end

**Comment by Jared-Fogle** [comment]
then people would just do it every time because thats funny

**Comment by necromanceranne** [comment]
This is pretty much ready by the by.

**Pull request merged**